### PR TITLE
multi pom support

### PR DESCRIPTION
**Pull request by jyasveer**

### CODE REVIEW

File: f8a_worker/workers/recommender.py [hunk 1]
```diff
@@ -434,32 +434,33 @@ class RecommendationTask(BaseTask):
...
-                "component_level": None,
-                }
-            }
-        else:
-            result = {"recommendations": {"similar_stacks": [], "component_level": None,}}
-
-        return result
+        for result in arguments['result']:
+            input_stack = {d["package"]: d["version"] for d in result.get("details", [])[0].get("_resolved")}
+            ecosystem = result.get("details", [])[0].get("ecosystem")
```
Comment (jpopelka) [inline f8a_worker/workers/recommender.py]: If there was no `'details'` in `result` you would see `IndexError: list index out of range` because you try to access its first ([0]) member.

If you change it to
```python
ecosystem = result["details"][0].get("ecosystem")
```
you would see `KeyError: 'details'`, which is more self-explanatory.

File: f8a_worker/data_normalizer.py [hunk 2]
```diff
@@ -61,7 +61,7 @@ def transform_keys(data, keymap):
                 in_key = in_k
             key = in_key if len(pair) == 1 else pair[1]
             out[key] = value
-
+
```
Comment (jpopelka) [inline f8a_worker/data_normalizer.py]: You are accidentally adding some white spaces here and on other places. Even they don't do any harm, it's a useless change.

File: f8a_worker/workers/graphaggregator.py [hunk 3]
```diff
@@ -52,16 +52,20 @@ def execute(self, arguments):
...
             # If we're handling an external request we need to convert dependency specifications to
             # concrete versions that we can query later on in the `AggregatorTask`
             manifest_descriptor = get_manifest_descriptor_by_filename(manifest['filename'])
             if 'external_request_id' in arguments:
+                manifest_dependencies = []
                 if manifest_descriptor.has_resolved_deps:  # npm-shrinkwrap.json, pom.xml, requirements.txt
                     if "_dependency_tree_lock" in out["details"][0]:  # npm-shrinkwrap.json, requirements.txt
-                        manifest_dependencies = out["details"][0]["_dependency_tree_lock"]["dependencies"]
+                        if 'dependencies' in out['details'][0]["_dependency_tree_lock"]:
+                            manifest_dependencies = out["details"][0]["_dependency_tree_lock"]["dependencies"]
```
Comment (jpopelka) [inline f8a_worker/workers/graphaggregator.py]: or
`manifest_dependencies = out["details"][0]["_dependency_tree_lock"].get("dependencies", [])`

File: f8a_worker/workers/graphaggregator.py [hunk 4]
```diff
@@ -52,16 +52,20 @@ def execute(self, arguments):
...
                     else:  # pom.xml
-                        manifest_dependencies = out["details"][0]["dependencies"]
+                        if 'dependencies' in out['details'][0]:
+                            manifest_dependencies = out["details"][0]["dependencies"]
```
Comment (jpopelka) [inline f8a_worker/workers/graphaggregator.py]: or
`manifest_dependencies = out["details"][0].get("dependencies", [])`

File: f8a_worker/workers/recommender.py (hunk 1 above)
Comment (jyasveer) [inline f8a_worker/workers/recommender.py]: Thanks for the comment. will change it

File: f8a_worker/workers/graphaggregator.py (hunk 3 above)
Comment (jyasveer) [inline f8a_worker/workers/graphaggregator.py]: do you recommend this way ?

Comment (jpopelka) [inline f8a_worker/workers/graphaggregator.py]: Both are fine, but mine is shorter, so yes, I recommend using `.get("dependencies", [])`

File: f8a_worker/workers/mercator.py [hunk 5]
```diff
@@ -194,12 +194,18 @@ def run_mercator(self, arguments, cache_path, keep_path=False, outermost_only=Tr
         result_data = {'status': 'unknown',
                        'summary': [],
                        'details': []}
-
         mercator_target = arguments.get('cache_sources_path', cache_path)
+        self.log.debug("############## mercator : run_mercator : arguments : start")
+        self.log.debug(arguments)
+        self.log.debug(mercator_target)
+        self.log.debug("############## mercator : run_mercator : arguments : end")
```
Comment (samuzzal-choudhury) [inline f8a_worker/workers/mercator.py]: Perhaps it was introduced for local debug. Please remove these lines

File: f8a_worker/workers/mercator.py [hunk 6]
```diff
@@ -194,12 +194,18 @@ def run_mercator(self, arguments, cache_path, keep_path=False, outermost_only=Tr
...
         tc = TimedCommand(['mercator', mercator_target])
         status, data, err = tc.run(timeout=timeout,
                                    is_json=True,
                                    update_env={'MERCATOR_JAVA_RESOLVE_POMS': 'true'})
+        self.log.debug("############## mercator : run_mercator : data : start")
+        self.log.debug(status, data, err)
+        self.log.debug("############## mercator : run_mercator : data : end")
```
Comment (samuzzal-choudhury) [inline f8a_worker/workers/mercator.py]: Perhaps it was introduced for local debug. Please remove these lines.

File: f8a_worker/data_normalizer.py (hunk 2 above)
Comment (samuzzal-choudhury) [inline f8a_worker/data_normalizer.py]: I agree with @jpopelka. Please, globally undo any unnecessary whitespace or lines additions.

File: f8a_worker/graphutils.py [hunk 7]
```diff
@@ -177,7 +177,8 @@ def aggregate_stack_data(stack, manifest_file, ecosystem):
             "total_licenses": len(stack_distinct_licenses),
             "distinct_licenses": list(stack_distinct_licenses),
             "popularity": stack_popularity_data,
-            "components": components
+            "components": components,
+            "manifest_file_path": manifest_file_path
```
Comment (samuzzal-choudhury) [inline f8a_worker/graphutils.py]: This is a schema change and can have widespread effect. So, we have to undergo discussions for such changes :-)

For now, please prefix the manifest_file_path with the manifest_name and provide that as the value for the "manifest_name" key.

File: f8a_worker/workers/graphaggregator.py [hunk 8]
```diff
@@ -52,16 +52,20 @@ def execute(self, arguments):
                                      .format(manifest['filename']))
             out["details"][0]['manifest_file'] = manifest['filename']
             out["details"][0]['ecosystem'] = manifest['ecosystem']
-
+            out["details"][0]['manifest_file_path'] = manifest['filepath']
```
Comment (samuzzal-choudhury) [inline f8a_worker/workers/graphaggregator.py]: please prefix the manifest_file_path with the manifest_name and set that new string as "manifest_name"

File: f8a_worker/workers/recommender.py [hunk 9]
```diff
@@ -434,32 +434,35 @@ class RecommendationTask(BaseTask):
...
-            result = {"recommendations": {
-                "similar_stacks": similarity_list,
-                "component_level": None,
-                }
-            }
-        else:
-            result = {"recommendations": {"similar_stacks": [], "component_level": None,}}
-
-        return result
+        for result in arguments['result']:
```
Comment (jpopelka) [inline f8a_worker/workers/recommender.py]: `for result in arguments.get('result', []):`

File: f8a_worker/workers/recommender.py [hunk 10]
```diff
@@ -434,32 +434,35 @@ class RecommendationTask(BaseTask):
...
+                # Calculate similarity of the filtered stacks
+                similar_stacks_list = rs.find_relative_similarity(input_stack, input_stack_vectors, ref_stacks)
+                similarity_list = self._get_stack_values(similar_stacks_list)
+                recommendations.append({
+                    "similar_stacks": similarity_list,
+                    "component_level": None,
+                    "manifest_file_path": manifest_file_path
+                    })
+            else:
+                recommendations.append({"similar_stacks": [], "component_level": None,})
```
Comment (jpopelka) [inline f8a_worker/workers/recommender.py]: Shouldn't there also be the "manifest_file_path" included ? i.e.
`recommendations.append({"similar_stacks": [], "component_level": None, "manifest_file_path": manifest_file_path})`

Comment (jyasveer) [inline f8a_worker/workers/recommender.py]: don't require here. if similar_stacks is empty we don't require this

File: f8a_worker/workers/mercator.py (hunk 5 above)
Comment (fridex) [inline f8a_worker/workers/mercator.py]: +1

File: f8a_worker/workers/mercator.py (hunk 6 above)
Comment (fridex) [inline f8a_worker/workers/mercator.py]: +1

File: f8a_worker/workers/graphaggregator.py (hunk 8 above)
Comment (msrb) [inline f8a_worker/workers/graphaggregator.py]: @samuzzal-choudhury Sam, are you OK with this?

**Comment by msrb** [comment]
Tested locally, seems to be working as expected. @jyasveer if you could remove the debug section as suggested by Sam and Frido, I can merge this PR.

Note the code is slowly becoming quite hard to read, but that's not a problem of this PR.

**Comment by msrb** [comment]
LGTM, thanks 😉
